File: Matrix1/Src/Interpolation.cpp
#include "Interpolation.hpp"
#include <cmath>
#include <stdexcept>

using namespace std;

Interpolation::Interpolation() {}

Interpolation::Interpolation(const vector<long double> &x, const vector<long double> &y)
{
    setData(x, y);
}

void Interpolation::setData(const vector<long double> &x, const vector<long double> &y)
{
    if (x.size() != y.size() || x.empty())
        throw invalid_argument("Interpolation data must be non-empty and x/y sizes must match.");

    // Ensure no duplicate x-values, because Lagrange basis would divide by zero.
    for (size_t i = 0; i < x.size(); ++i)
        for (size_t j = i + 1; j < x.size(); ++j)
            if (fabsl(x[i] - x[j]) < 1e-18L)
                throw invalid_argument("Interpolation x-values must be unique.");

    // Store interpolation points directly in inherited Matrix storage as n x 2.
    rows = static_cast<int>(x.size());
    cols = 2;
    mat.assign(static_cast<size_t>(rows), vector<long double>(static_cast<size_t>(cols), 0.0L));
    for (int i = 0; i < rows; ++i)
    {
        mat[static_cast<size_t>(i)][0] = x[static_cast<size_t>(i)];
        mat[static_cast<size_t>(i)][1] = y[static_cast<size_t>(i)];
    }
}

int Interpolation::pointCount() const
{
    return rows;
}

vector<long double> Interpolation::getX() const
{
    vector<long double> x;
    x.reserve(static_cast<size_t>(rows));
    for (int i = 0; i < rows; ++i)
        x.push_back(mat[static_cast<size_t>(i)][0]);
    return x;
}

vector<long double> Interpolation::getY() const
{
    vector<long double> y;
    y.reserve(static_cast<size_t>(rows));
    for (int i = 0; i < rows; ++i)
        y.push_back(mat[static_cast<size_t>(i)][1]);
    return y;
}

Lagrange::Lagrange() : Interpolation() {}

Lagrange::Lagrange(const vector<long double> &x, const vector<long double> &y) : Interpolation(x, y) {}

long double Lagrange::evaluate(long double x) const
{
    if (rows <= 0 || cols < 2 || mat.empty())
        throw runtime_error("No interpolation data is set.");

    long double result = 0.0L;

    for (int i = 0; i < rows; ++i)
    {
        long double basis = 1.0L;

        for (int j = 0; j < rows; ++j)
        {
            if (i == j)
                continue;

            const long double den = mat[static_cast<size_t>(i)][0] - mat[static_cast<size_t>(j)][0];
            if (fabsl(den) < 1e-18L)
                throw runtime_error("Duplicate x-values found during evaluation.");

            basis *= (x - mat[static_cast<size_t>(j)][0]) / den;
        }

        result += mat[static_cast<size_t>(i)][1] * basis;
    }

    return result;
}

vector<pair<long double, long double>> Lagrange::evaluateRange(long double xMin, long double xMax, int samples) const
{
    if (samples < 2)
        throw invalid_argument("samples must be >= 2");

    vector<pair<long double, long double>> curve;
    curve.reserve(static_cast<size_t>(samples));

    const long double step = (xMax - xMin) / static_cast<long double>(samples - 1);
    for (int i = 0; i < samples; ++i)
    {
        const long double x = xMin + static_cast<long double>(i) * step;
        curve.push_back({x, evaluate(x)});
    }

    return curve;
}

File: Matrixpy/Include/interpolation.py
from Matrixpy.Include.matrix import Matrix

class Interpolation(Matrix):
    def __init__(self, x=None, y=None):
        if x is not None and y is not None:
            super().__init__(len(x), 2)
            self.setData(x, y)
        else:
            super().__init__(0, 0)

    def setData(self, x, y):
        from Matrixpy.Src.Interpolation import set_data
        set_data(self, x, y)

    def pointCount(self):
        return self.rows

    def getX(self):
        return [self.mat[i][0] for i in range(self.rows)]

    def getY(self):
        return [self.mat[i][1] for i in range(self.rows)]

    def evaluate(self, x):
        raise NotImplementedError("Subclasses must implement evaluate()")

    def evaluateRange(self, xMin, xMax, samples):
        raise NotImplementedError("Subclasses must implement evaluateRange()")

class Lagrange(Interpolation):
    def __init__(self, x=None, y=None):
        super().__init__(x, y)

    def evaluate(self, x):
        from Matrixpy.Src.Interpolation import evaluate_lagrange
        return evaluate_lagrange(self, x)

    def evaluateRange(self, xMin, xMax, samples):
        from Matrixpy.Src.Interpolation import evaluate_range_lagrange
        return evaluate_range_lagrange(self, xMin, xMax, samples)
